platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/http/HttpRequest.java

```diff
@@ -0,0 +1,95 @@
+package com.mapbox.mapboxsdk.http;
+
+import android.content.Context;
+import android.content.pm.PackageInfo;
+import android.support.annotation.NonNull;
+import com.mapbox.mapboxsdk.Mapbox;
+
+/**
+ * Base class for performing core http requests.
+ * <p>
+ * This abstraction allows to provide alternative implementations for the http interaction of this library.
+ * </p>
+ */
+public abstract class HttpRequest {
+
+  protected static final int CONNECTION_ERROR = 0;
+  protected static final int TEMPORARY_ERROR = 1;
+  protected static final int PERMANENT_ERROR = 2;
+
+  /**
+   * Executes the request.
+   *
+   * @param httpRequest callback to be invoked when we receive a response
+   * @param nativePtr   the pointer associated to the request
+   * @param resourceUrl the resource url to download
+   * @param etag        http header, identifier for a specific version of a resource
+   * @param modified    http header, used to determine if a resource hasn't been modified since
+   */
+  public abstract void executeRequest(HttpRequestResponder httpRequest, long nativePtr, String resourceUrl,
```

Comment:
composition is provided in the module file [here](https://github.com/mapbox/mapbox-gl-native/pull/12468/files/b73d3cad75c1700c0f7d4ea6604f49eeaa2bef9b#diff-b5f1ea293c8bb5c15d761f065f438fb6R10). The HttpRequest class is the contract between the SDK and the alternate implementation. How would you enforce the availability of predefined methods, the sdk needs a predefined hook, either an interface or an abstract class?

Reply:
An interface is generally preferable since it doesn't lock the API user into single inheritance. I think an interface should work here unless I'm missing something.